stan/math/gpu/kernel_cl.hpp

```diff
@@ -29,24 +29,46 @@ map_base_opts base_opts
        {"UPPER", static_cast<int>(TriangularViewGPU::Upper)},
        {"ENTIRE", static_cast<int>(TriangularViewGPU::Entire)},
        {"UPPER_TO_LOWER", static_cast<int>(TriangularMapGPU::UpperToLower)},
-       {"LOWER_TO_UPPER", static_cast<int>(TriangularMapGPU::LowerToUpper)}};
+       {"LOWER_TO_UPPER", static_cast<int>(TriangularMapGPU::LowerToUpper)},
+       {"WORK_PER_WI_MULT", 8},
+       {"WG_SIZE_MULT", 32},
+       {"WG_SIZE_MULT_SELF_TRANS", 32},
+       {"WORK_PER_WI_MULT_SELF_TRANS", 4}};
 
 auto compile_kernel(const char* name, const char* source) {
+  size_t max_wg_size = opencl_context.max_workgroup_size();
+  int wg_size_sqrt = static_cast<int>(sqrt(static_cast<double>(max_wg_size)));
+  // Does a compile time check of the maximum allowed
+  // dimension of a square workgroup size
+  // WG size of (32,32) works on all recent GPU but would fail on some
+  // older integrated GPUs or CPUs
+  if (wg_size_sqrt < base_opts["WG_SIZE_MULT"]){
+    base_opts["WG_SIZE_MULT"] = wg_size_sqrt;
+    base_opts["WORK_PER_WI_MULT"] = 1;
+  }
+  if (wg_size_sqrt < base_opts["WG_SIZE_MULT_SELF_TRANS"]){
+    base_opts["WG_SIZE_MULT_SELF_TRANS"] = wg_size_sqrt;
+    base_opts["WORK_PER_WI_MULT_SELF_TRANS"] = 1;
+  }
   std::string kernel_opts = "";
   for (auto&& comp_opts : base_opts) {
     kernel_opts += std::string(" -D") + comp_opts.first + "="
                    + std::to_string(comp_opts.second);
   }
   std::string kernel_source(opencl_kernels::helpers);
   kernel_source.append(source);
+  cl::Program program;
   try {
     cl::Program::Sources src(1, std::make_pair(kernel_source.c_str(),
                                                strlen(kernel_source.c_str())));
-    cl::Program program = cl::Program(opencl_context.context(), src);
+    program = cl::Program(opencl_context.context(), src);
     program.build({opencl_context.device()}, kernel_opts.c_str());
 
     return cl::Kernel(program, name);
   } catch (const cl::Error& e) {
+    std::string buildlog = program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(
+        opencl_context.device()[0]);
+    std::cerr << "Build log :" << std::endl << buildlog << std::endl;
```

Comment:
Writing to std::cerr is probably not OK. Where can I put this? I would really like to output the logs in case of a kernel compile error. Just giving out error -11 is not really informative.

Reply:
I think we normally just throw exceptions as our primary way of communication on error conditions. Can you wrap up the build log into an exception to throw?